plutus-core/plutus-core/src/Universe/Core.hs

```diff
@@ -0,0 +1,756 @@
+type Contains :: forall k. (Type -> Type) -> k -> Constraint
+class uni `Contains` a where
+    knownUni :: uni (T a)
+
+{- Note [The definition of Includes]
+We need to be able to partially apply 'Includes' (required in the definition of '<:' for example),
+however if we define 'Includes' as a class alias like that:
+
+    class    Contains uni `Permits` a => uni `Includes` a
+    instance Contains uni `Permits` a => uni `Includes` a
+
+we get this extra annoying warning:
+
+    • The constraint ‘Includes uni ()’ matches
+        instance forall k (uni :: * -> *) (a :: k).
+                 Permits (Contains uni) a =>
+                 Includes uni a
+      This makes type inference for inner bindings fragile;
+        either use MonoLocalBinds, or simplify it using the instance
+
+at the use site, so instead we define 'Includes' as a type alias of one argument (i.e. 'Includes'
+has to be immediately applied only to a @uni@ at the use site).
+-}
+
+-- See Note [The definition of Includes].
+-- | @uni `Includes` a@ reads as \"@a@ is in the @uni@\". @a@ can be of a higher-kind,
+-- see the docs of 'Contains' on why you might want that.
+type Includes :: forall k. (Type -> Type) -> k -> Constraint
+type Includes uni = Permits (Contains uni)
+
+-- | Same as 'knownUni', but receives a @proxy@.
+knownUniOf :: uni `Contains` a => proxy a -> uni (T a)
+knownUniOf _ = knownUni
+
+-- | Wrap a value into @Some (ValueOf uni)@, given its explicit type tag.
+someValueOf :: forall a uni. uni (T a) -> a -> Some (ValueOf uni)
+someValueOf uni = Some . ValueOf uni
+
+-- | Wrap a value into @Some (ValueOf uni)@, provided its type is in the universe.
+someValue :: forall a uni. uni `Includes` a => a -> Some (ValueOf uni)
+someValue = someValueOf knownUni
+
+-- | A monad to decode types from a universe in.
+newtype DecodeUniM a = DecodeUniM
+    { unDecodeUniM :: StateT [Int] Maybe a
+    } deriving newtype (Functor, Applicative, Alternative, Monad, MonadPlus, MonadFail)
+
+runDecodeUniM :: [Int] -> DecodeUniM a -> Maybe (a, [Int])
+runDecodeUniM is (DecodeUniM a) = runStateT a is
+
+-- | A universe is 'Closed', if it's known how to constrain every type from the universe.
+-- The universe doesn't have to be finite and providing support for infinite universes is the
+-- reason why we encode a type as a sequence of integer tags as opposed to a single integer tag.
+-- For example, given
+--
+-- >   data U a where
+-- >       UList :: !(U a) -> U [a]
+-- >       UInt  :: U Int
+--
+-- @UList (UList UInt)@ can be encoded to @[0,0,1]@ where @0@ and @1@ are the integer tags of the
+-- @UList@ and @UInt@ constructors, respectively.
+class Closed uni where
+    -- | A constrant for \"@constr a@ holds for any @a@ from @uni@\".
+    type Everywhere uni (constr :: Type -> Constraint) :: Constraint
+
+    -- | Encode a type as a sequence of 'Int' tags.
+    -- The opposite of 'decodeUni'.
+    encodeUni :: uni a -> [Int]
```

Comment:
This encodes a type in a universe, not the universe itself, so maybe it should be called something else? In fact, isn't this encoding the _tag_?

Reply:
Also, why *is* encoding/decoding in `Closed`? The description of `Closed` is all about `Everywhere`...